File: 07 - More Operation on Single Linked List/7_4 Delete at any position/go/deleteAtAnyPosition.go
package main

import "fmt"

type Node struct{
	val int
	next *Node
}

func insertAtTail(head **Node, tail **Node, value int){
	newNode := &Node{val: value}
	
	if *head == nil{
		*head = newNode;
		*tail = newNode
		return;
	}
	(*tail).next = newNode;
	*tail = newNode;
}

func printLinkedList(head *Node){
	temp := head;
	for {
		if temp == nil{
			break;
		}else{
			fmt.Println(temp.val);
			temp = temp.next;
		}
	}
}

func deleteAtAnyPosition(head **Node, idx int){
	tmp := *head;
	for i:=1; i < idx; i++{
		tmp = tmp.next
	}
	deletedNoted := tmp.next;
	tmp.next = deletedNoted.next;
}

func main(){

	var head *Node = nil;
	var tail *Node = nil

	var value int;
	for {
		fmt.Scan(&value)
		if value == -1{
			break;
		}
		insertAtTail(&head, &tail, value)
	}
	deleteAtAnyPosition(&head, 3);
	printLinkedList(head)
}

File: 09 - Doubly LinkedList/9_8 Delete at any position/deleteAtAnyPosition.cpp
#include<bits/stdc++.h>
using namespace std;

class Node {
    public:
    int val;
    Node* next;
    Node* prev;
    Node(int val) {
        this->val = val;
        this->prev = NULL;
        this->next =NULL;
    }
};

void deleteAtAnyPosition(Node* &head, Node* &tail, int pos){
    if(head == NULL) return;
    Node* tmp = head;
    for(int i = 0; i < pos-1; i++){
        tmp = tmp->next;
    }
    Node* deletedNode = tmp->next;
    tmp->next = tmp->next->next;
    tmp->next->prev = tmp;
    delete deletedNode;
}

void printForward(Node* head){
    Node* tmp = head;
    while (tmp != NULL)
    {
        /* code */
        cout << tmp->val << " ";
        tmp = tmp->next;
    }
    cout << endl;
    
}

void printBackward(Node* tail){
    Node* tmp = tail;
    while(tmp!=NULL){
        cout << tmp->val << " ";
        tmp = tmp->prev;
    }
    cout << endl;
}

int main()
{
    Node* head = new Node(10);
    Node* a = new Node(20);
    Node* tail = new Node(30);

    head->next = a;
    a->prev = head;
    a->next = tail;
    tail->prev = a;

    printForward(head);
    deleteAtAnyPosition(head, tail, 1);
    printForward(head);

    return 0;
}